drivers/charger/sbs_charger.c

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright 2023 Cirrus Logic, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT sbs_sbs_charger
+
+#include <zephyr/drivers/charger.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+
+#include "sbs_charger.h"
+
+struct sbs_charger_config {
+	struct i2c_dt_spec i2c;
+};
+
+LOG_MODULE_REGISTER(sbs_charger);
+
+static int sbs_cmd_reg_read(const struct device *dev, uint8_t reg_addr, uint16_t *val)
+{
+	const struct sbs_charger_config *cfg;
+	uint8_t i2c_data[2];
+	int status;
+
+	cfg = dev->config;
+	status = i2c_burst_read_dt(&cfg->i2c, reg_addr, i2c_data, sizeof(i2c_data));
+	if (status < 0) {
+		LOG_ERR("Unable to read register");
+		return status;
+	}
+
+	*val = sys_get_le16(i2c_data);
+
+	return 0;
+}
+
+static int sbs_cmd_reg_write(const struct device *dev, uint8_t reg_addr, uint16_t val)
+{
+	const struct sbs_charger_config *config = dev->config;
+	uint8_t buf[2];
+
+	sys_put_le16(val, buf);
+
+	return i2c_burst_write_dt(&config->i2c, reg_addr, buf, sizeof(buf));
+}
+
+static int sbs_cmd_reg_update(const struct device *dev, uint8_t reg_addr, uint16_t mask,
+			      uint16_t val)
+{
+	uint16_t old_val, new_val;
+	int ret;
+
+	ret = sbs_cmd_reg_read(dev, SBS_CHARGER_REG_STATUS, &old_val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	new_val = (old_val & ~mask) | (val & mask);
+	if (new_val == old_val) {
+		return 0;
+	}
+
+	return sbs_cmd_reg_write(dev, reg_addr, new_val);
+}
+
+static int sbs_charger_get_prop(const struct device *dev, const charger_prop_t prop,
+				union charger_propval *val)
+{
+	uint16_t reg_val;
+	int ret;
+
+	switch (prop) {
+	case CHARGER_PROP_ONLINE:
+		ret = sbs_cmd_reg_read(dev, SBS_CHARGER_REG_STATUS, &reg_val);
+		if (reg_val & SBS_CHARGER_STATUS_AC_PRESENT) {
```

Comment:
if the command above fails you should not use `reg_val`, not that it's going to hurt anything but you'd better off adding few `if (ret < 0) { return ret; }` and be more defensive about the whole story. I would also change the break below here to `return 0;`, saves you from thinking "is the rest of the function doing something else?"

Reply:
ack